go-chaos/internal/cluster.go

```diff
@@ -0,0 +1,120 @@
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+	"strings"
+	"time"
+)
+
+func (c K8Client) GetZeebeStatefulSet() (*v1.StatefulSet, error) {
+	namespace := c.GetCurrentNamespace()
+	ctx := context.TODO()
+
+	helmLabel := meta.LabelSelector{
+		MatchLabels: map[string]string{"app.kubernetes.io/name": "zeebe"},
+	}
+
+	statefulSets := c.Clientset.AppsV1().StatefulSets(namespace)
+	sfs, err := statefulSets.List(ctx, meta.ListOptions{LabelSelector: labels.Set(helmLabel.MatchLabels).String()})
+	if err != nil {
+		return nil, err
+	}
+	if len(sfs.Items) == 1 {
+		return &sfs.Items[0], nil
+	}
+	if len(sfs.Items) == 0 {
+		// On SaaS the StatefulSet is just named "zeebe" without any identifying labels
+		return statefulSets.Get(ctx, "zeebe", meta.GetOptions{})
+	}
+	return nil, errors.New("could not uniquely identify the stateful set for Zeebe")
+}
+
+// ScaleZeebeCluster Scales the StatefulSet for Zeebe. Waits until scaling is complete before returning the initial scale.
+func (c K8Client) ScaleZeebeCluster(replicas int) (int, error) {
+	namespace := c.GetCurrentNamespace()
+	ctx := context.TODO()
+	var initialReplicas int
+
+	sfs, err := c.GetZeebeStatefulSet()
+	if err != nil {
+		return 0, err
+	}
+	statefulSets := c.Clientset.AppsV1().StatefulSets(namespace)
+	err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		currentScale, err := statefulSets.GetScale(ctx, sfs.Name, meta.GetOptions{})
+		initialReplicas = int(currentScale.Spec.Replicas)
+		if err != nil {
+			return err
+		}
+		currentScale.Spec.Replicas = int32(replicas)
+		_, err = statefulSets.UpdateScale(ctx, sfs.Name, currentScale, meta.UpdateOptions{})
+		return err
+	})
+	if err != nil {
+		return 0, err
+	}
+	err = wait.PollImmediateUntilWithContext(
+		ctx,
+		1*time.Second,
+		func(ctx context.Context) (done bool, err error) {
+			scale, err := c.Clientset.AppsV1().StatefulSets(namespace).GetScale(ctx, sfs.Name, meta.GetOptions{})
+			if err != nil {
+				return false, err
+			}
+			fmt.Printf("Waiting for %d replicas, currently at %d \n", replicas, scale.Status.Replicas)
+			return scale.Status.Replicas == int32(replicas), nil
+		},
+	)
+
+	return initialReplicas, err
+}
+
+func (c K8Client) PauseReconciliation() error {
+	return c.setPauseFlag(true)
+}
+
+func (c K8Client) ResumeReconciliation() error {
+	return c.setPauseFlag(false)
+}
+
+func (c K8Client) setPauseFlag(pauseEnabled bool) error {
+	ctx := context.TODO()
+	namespace := c.GetCurrentNamespace()
+	clusterId := strings.TrimSuffix(namespace, "-zeebe")
+	zeebeCrd := schema.GroupVersionResource{Group: "cloud.camunda.io", Version: "v1alpha1", Resource: "zeebeclusters"}
+	payload := fmt.Sprintf(`{"metadata": {"labels": {"cloud.camunda.io/pauseReconciliation": "%t"}}}`, pauseEnabled)
```

Comment:
TIL :bulb: 

Interesting :+1: previously we had to update the CRD

Reply:
That's what this is. It uses the dynamic client to get the custom resource for the zeebe cluster and patches it with a new label.

Reply:
This label is new, before it was part of the crd